File: studio/AuraStudio/Sources/AuraStudio/Models/FirmwareFamily.swift
import Foundation

/// Que firmware de la familia Aura dice ser el que esta instalado, leido
/// de la clave `firmware_family` de `.rockbox/aura/aura.cfg` (contrato
/// v8, ST-046).
///
/// Es un TERCER hecho, deliberadamente separado de los dos que ya
/// distingue `AuraDevice` (ST-016):
///
///   1. `AuraDevice.firmware` -- que ARCHIVOS hay en el disco.
///   2. `AuraDevice.runningFirmware` -- quien atiende el USB ahora.
///   3. `declaredFamily` (esto) -- que dice el firmware DE SI MISMO.
///
/// Hacia falta un tercero porque los otros dos no pueden responder la
/// pregunta: Metro-Aura escribe en `.rockbox/aura/` exactamente igual que
/// Aura (comparten el contrato entero de §D a proposito), y por USB los
/// dos se anuncian como "Rockbox.org" -- son forks del mismo Rockbox
/// (`USBDeviceIdentity.classify`). Sin una clave declarada no hay forma
/// de distinguirlos, y Studio trataba a Metro como Aura: le ofrecia
/// actualizaciones del repo equivocado, que al aceptarse lo habrian
/// SOBRESCRITO.
///
/// **La ausencia de la clave significa `.aura`**, y eso es lo que hace el
/// cambio retrocompatible: Aura-Firmware nunca la escribio ni la
/// escribira, asi que todo iPod con Aura instalada -- incluidos los
/// instalados antes de esta version de Studio -- cae en el caso correcto
/// sin tocar el firmware. Metro si la escribe (`metro_settings.c`, M-004).
enum FirmwareFamily: Equatable {
    case aura
    case metro
    /// Una familia que esta version de Studio no conoce. Se conserva el
    /// texto crudo para poder mostrarlo y para no fingir que es Aura: un
    /// firmware que se molesto en declararse NO es Aura, aunque no
    /// sepamos cual es.
    case unknown(String)

    /// Valor tal como aparece en `aura.cfg`. `nil` para Aura: la clave
    /// simplemente no existe (ver arriba).
    var configValue: String? {
        switch self {
        case .aura: return nil
        case .metro: return "metro"
        case .unknown(let raw): return raw
        }
    }

    /// Nombre de producto, para la UI.
    var displayName: String {
        switch self {
        case .aura: return "Aura"
        case .metro: return "Metro"
        case .unknown(let raw): return raw
        }
    }

    /// Repositorio de GitHub que publica los Releases de esta familia
    /// (`owner/repo`). `nil` para una familia desconocida: no hay a donde
    /// preguntar, y adivinar seria peor que no ofrecer actualizaciones.
    var releaseRepository: String? {
        switch self {
        case .aura: return "Ricolinos/Aura-Firmware"
        case .metro: return "Ricolinos/Metro-Aura"
        case .unknown: return nil
        }
    }

    /// ST-047: las dos familias que esta version de Studio trae
    /// EMBEBIDAS y por lo tanto puede instalar. Una familia desconocida
    /// se detecta pero no se instala.
    static let installable: [FirmwareFamily] = [.aura, .metro]

    var isInstallable: Bool { Self.installable.contains(self) }

    /// Subdirectorio de Resources donde viven los artefactos de la
    /// familia (ver project.yml / scripts/fetch-firmware.sh). `nil` =
    /// raiz del bundle, que es donde Aura siempre estuvo: mover los
    /// suyos habria sido riesgo sin beneficio.
    var bundleSubdirectory: String? {
        switch self {
        case .aura: return nil
        case .metro: return "metro"
        case .unknown: return nil
        }
    }

    /// Un archivo del arbol `.rockbox/` que el firmware carga al arrancar
    /// -- el instalador lo usa como centinela de "el zip se extrajo
    /// completo" (InstallerViewModel.copyFirmwareFiles). Cada familia
    /// trae sus propias fuentes, asi que el centinela es por familia.
    var installedTreeSentinel: String? {
        switch self {
        case .aura: return ".rockbox/fonts/a26-title-20.fnt"
        case .metro: return ".rockbox/fonts/metro-list-20.fnt"
        case .unknown: return nil
        }
    }

    /// URL publica del repositorio (para Licencias y para "ver el
    /// Release").
    var repositoryURL: URL? {
        guard let repo = releaseRepository else { return nil }
        return URL(string: "https://github.com/\(repo)")
    }

    /// ST-056 / contrato v10: nombre del arbol DORMIDO de esta familia en
    /// la raiz del iPod (`/.firmware-aura/`, `/.firmware-metro/`): un
    /// `.rockbox` completo, en reposo, con sus propios ajustes. El activo
    /// es siempre `/.rockbox/` (lo unico que el bootloader arranca).
    var dormantTreeName: String? {
        switch self {
        case .aura: return ".firmware-aura"
        case .metro: return ".firmware-metro"
        case .unknown: return nil
        }
    }

    /// Interpreta el valor crudo de la clave. Insensible a mayusculas y
    /// espacios porque el parser del firmware (`settings_parseline()`) no
    /// normaliza nada: lo que se escriba es lo que se lee.
    static func parse(configValue raw: String?) -> FirmwareFamily {
        guard let raw else { return .aura }
        let value = raw.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        switch value {
        case "": return .aura
        case "aura": return .aura
        case "metro": return .metro
        default: return .unknown(value)
        }
    }
}

File: studio/windows/AuraStudio.Core/FirmwareFamily.cs
namespace AuraStudio.Core;

/// <summary>
/// Qué familia de firmware dice ser la que está instalada, leída de la
/// clave `firmware_family` de `.rockbox/aura/aura.cfg` (contrato v8, ST-046).
///
/// Es un TERCER hecho, deliberadamente separado de los dos que ya distingue
/// <c>USBDeviceIdentity</c>/<c>RunningFirmware</c> (ST-016):
/// 1) qué ARCHIVOS hay en el disco, 2) quién atiende el USB ahora, y 3) qué
/// dice el firmware DE SÍ MISMO. Hace falta un tercero porque Metro-Aura
/// escribe en `.rockbox/aura/` exactamente igual que Aura y por USB ambos se
/// anuncian como "Rockbox.org" — sin una clave declarada Studio trataría a
/// Metro como Aura.
///
/// **La ausencia de la clave significa Aura** y eso hace el cambio
/// retrocompatible: todo iPod con Aura (incluidos los instalados antes de
/// esta versión de Studio) cae en el caso correcto sin tocar el firmware.
///
/// El Swift original es un <c>enum</c> con valor asociado para la familia
/// desconocida (<c>unknown(String)</c>, que conserva el texto crudo). Por eso
/// acá no puede ser un <c>enum</c> plano de C#: se modela como una clase con
/// instancias singleton, el mismo patrón que <c>RunningFirmware</c> ya usa en
/// este proyecto. Una familia desconocida se crea con <see cref="Unknown"/>.
/// </summary>
public sealed class FirmwareFamily : IEquatable<FirmwareFamily>
{
    /// <summary>Firmware original de Aura. Su firma es la AUSENCIA de la clave (ST-046).</summary>
    public static FirmwareFamily Aura { get; } = new("aura", null, "Aura", "Ricolinos/Aura-Firmware", ".rockbox/fonts/a26-title-20.fnt");

    /// <summary>Metro-Aura (`Ricolinos/Metro-Aura`, M-004).</summary>
    public static FirmwareFamily Metro { get; } = new("metro", "metro", "Metro", "Ricolinos/Metro-Aura", ".rockbox/fonts/metro-list-20.fnt");

    /// <summary>ST-065: tercera familia, moonlit.aura (`Ricolinos/moonlit-aura`).</summary>
    public static FirmwareFamily Moonlit { get; } = new("moonlit", "moonlit", "moonlit.aura", "Ricolinos/moonlit-aura", ".rockbox/fonts/moonlit-body-18.fnt");

    /// <summary>
    /// Las familias que esta versión de Studio trae EMBEBIDAS y por lo tanto
    /// puede instalar (ST-047/ST-065). Una familia desconocida se detecta pero
    /// no se instala. Es la ÚNICA lista de familias: todo lo que las enumera
    /// itera sobre esto.
    /// </summary>
    public static readonly IReadOnlyList<FirmwareFamily> Installable = new[] { Aura, Metro, Moonlit };

    private readonly string _raw;

    /// <summary>Valor tal como aparece en `aura.cfg`. `null` para Aura: la clave simplemente no existe.</summary>
    public string? ConfigValue { get; }

    /// <summary>Nombre de producto, para la UI.</summary>
    public string DisplayName { get; }

    /// <summary>
    /// Repositorio de GitHub (`owner/repo`) donde esta familia publica sus
    /// Releases. `null` para una familia desconocida: sin repo no hay a
    /// donde preguntar (ST-046).
    /// </summary>
    public string? ReleaseRepository { get; }

    /// <summary>
    /// Un archivo del árbol `.rockbox/` que el firmware carga al arrancar —
    /// el instalador lo usa como centinela de "el zip se extrajo completo".
    /// Cada familia trae sus propias fuentes, así que el centinela es por
    /// familia. `null` para Aura/unknown (Aura se reconoce por ausencia, no
    /// por centinela).
    /// </summary>
    public string? InstalledTreeSentinel { get; }

    public bool IsInstallable => Installable.Contains(this);

    /// <summary>
    /// Nombre del árbol dormido de esta familia en la raíz del volumen
    /// (`/.firmware-aura/`, `/.firmware-metro/`, `/.firmware-moonlit/` —
    /// contrato v10/v14 §A bis). `null` para una familia desconocida: no se
    /// estaciona ni se despierta lo que no se sabe qué es.
    /// </summary>
    public string? DormantTreeName => IsInstallable ? $".firmware-{_raw}" : null;

    private FirmwareFamily(string raw, string? configValue, string displayName, string? releaseRepository, string? sentinel)
    {
        _raw = raw;
        ConfigValue = configValue;
        DisplayName = displayName;
        ReleaseRepository = releaseRepository;
        InstalledTreeSentinel = sentinel;
    }

    /// <summary>Una familia que esta versión de Studio no conoce. Conserva el texto crudo para poder mostrarlo y para no fingir que es Aura.</summary>
    public static FirmwareFamily Unknown(string raw) => new(raw, raw, raw, null, null);

    /// <summary>
    /// Interpreta el valor crudo de la clave. Insensible a mayúsculas y
    /// espacios porque el parser del firmware (`settings_parseline()`) no
    /// normaliza nada: lo que se escriba es lo que se lee. `null`/vacío =
    /// Aura.
    /// </summary>
    public static FirmwareFamily Parse(string? raw)
    {
        if (raw is null) return Aura;
        string value = raw.Trim().ToLowerInvariant();
        return value switch
        {
            "" => Aura,
            "aura" => Aura,
            "metro" => Metro,
            "moonlit" => Moonlit,
            _ => Unknown(value),
        };
    }

    public bool Equals(FirmwareFamily? other) => other is not null && _raw == other._raw;

    public override bool Equals(object? obj) => obj is FirmwareFamily other && Equals(other);

    public override int GetHashCode() => _raw.GetHashCode();

    public override string ToString() => DisplayName;
}
